### ISSUE TITLE
[PATCH] Add --umask option to twistd

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10647]
|<img alt="thomasvs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| thomasvs reported|
|-|-|
|Trac ID|trac#966|
|Type|defect|
|Created|2005-04-05 21:20:33Z|
|Branch|https://github.com/twisted/twisted/tree/twistd-umask-966|

Attachments:

* [umask.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bdf/bdf59a109a30762380d851c187b7ec5256d89b72/61b20fd7e23e1128dad94c6bb1bb317a4b1350ea.patch) (6985 bytes) - added by thomasvs on 2005-04-05 21:20:30Z - 
* [umask.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bdf/bdf59a109a30762380d851c187b7ec5256d89b72/e5c7eb72ed5b460cdbf188096dd6c4ca9136ad13.patch) (7044 bytes) - added by thomasvs on 2005-04-05 21:57:52Z - 
* [umask.4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bdf/bdf59a109a30762380d851c187b7ec5256d89b72/cd4ecabd5cc5128aa7dca61dda928d6a4ae9dfe4.patch) (3693 bytes) - added by zectbumo on 2008-02-09 22:28:17Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__966 966
type__defect defect
reporter__thomasvs thomasvs
priority__normal normal
milestone__ 
branch__branches_twistd_umask_966 branches/twistd-umask-966
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1112736033000000 1112736033000000
changetime__1217110893000000 1217110893000000
version__ 
owner__ 
cc__exarkun cc__spiv cc__thomasvs cc__zectbumo
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="thomasvs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|thomasvs commented|
|-|-|

```
#!html
<pre>
... which makes all files created under a twistd daemon (say, e.g., buildbot)
only readable by user.

As mentioned in http://www.erlenstar.demon.co.uk/unix/faq_2.html#SEC16 and the
Stevens UNIX book, a daemon should default to doing umask(0) instead.

Attached patch provides:
- a --umask option for twistd
- a umask ivar on Process
- changing the umask for both daemon and non-daemon if set with --umask
- leaving umask alone if not a daemon
- setting umask to 0 if daemon
- add to tests
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="thomasvs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|thomasvs commented|
|-|-|

```
#!html
<pre>
and btw, I tested all combinations I could think of by making chatserver.py
touch a file on new connections :)
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="thomasvs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|thomasvs commented|
|-|-|

```
#!html
<pre>
better patch - only do the int() on the command line arg, because aProcess.umask
would not like it very much.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
I'd prefer it if it only accepted octal on the command line, same as chmod(1)
does.  Otherwise, e.g. "--umask 222" will surprise people.

I'm also concerned by the backwards compatibility problem here... existing users
of twistd might rely on the current umask of 0077, so we should probably leave
the default as that.  I don't see much harm in being "secure by default".  Or at
the very least, the patch should update twisted/topfiles/NEWS to warn people
upgrading that they may need to set a umask explicitly.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="thomasvs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|thomasvs commented|
|-|-|

```
#!html
<pre>
ok, I'll change the patch to accept only octal.  (btw, chmod accepts strings as
well for the mode, just like my bash umask does :)) Would you prefer it silently
accepting --umask 222, or force the user to do --umask 0222 so that they don't
forget it's octal ?

I agree with the backwards compat as well.  However, imo this only applies until
twisted 2.1, no ? Saying "I don't see much harm in being secure by default"
really is a way of saying "I do see why it's wrong but I don't want the pain of
this change if there is pain."  The code references a certain page of accepted
practices for all of its decisions, and this one is on that page :)
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Personally, I'd be happy with it accepting both "0222" and "222" on the command
line, which is what int(foo, 8) will accept.

Certainly, for 2.1 we can consider changing the default.  But, I do actually
think the existing default is reasonable: if I don't think about what
permissions I expect files created by my daemon to have, I may as well make them
unreadable except by me.  It should be easily overridden, and that's what a
--umask flag to twistd will provide.

Another thing to consider is that I expect twistd currently relies on the umask
of 0077 when writing pid files, log files, and shutdown taps.  The shutdown tap
in particular is may be a security concern; it certainly shouldn't be writeable
by world by default!  Log files might contain sensitive information too.

Anyway, I'm basically happy with the patch as proposed, but I think some other
Twisted dev should review it before committing... I'm hardly the last word on
twistd :)
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Updating title.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="zectbumo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|zectbumo commented|
|-|-|

```
#!html
<pre>
The FAQ says umask is optional.  Which in that case the umask should be left
alone.  After the thomasvs patch, if a user wants to change it they can supply
the --umask option.  If the thomasvs fix didn't "force" a umask(0) then a user
can simply set the umask before running the daemon.  When reading specs, one
needs to pay attention to "should" and "can".  Does the Stevens book say
"should" or "can?"  The FAQ says "can," so is there a conflict here?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

This still looks like a good idea to me.  Adding review keyword -- there's a patch, it has tests.

**Comment by twisted-trac** [comment]
|<img alt="zectbumo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|zectbumo commented|
|-|-|

I got a copy of Unix Network Programming by Stevens and it says "A daemon should execute umask(0); to reset this mask."
I also found this great example (also doing a umask(0)).
See createDaemon function:
http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/278731/index_txt

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

Review tickets should be "highest" priority so they don't languish.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

The test coverage for this is a bit lacking.  It doesn't actually cover the feature being changed.  I'll probably try to write a test for this at some point, but I'd be happy if someone beat me to it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

[#3024](https://github.com/twisted/twisted/issues/3024) has something to do with this.

**Comment by twisted-trac** [comment]
|<img alt="zectbumo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|zectbumo commented|
|-|-|

I added umask.4.patch so it patches the current twisted code and changed int(umask, 0) to int(umask, 8).  I left out the unit test since exarkun or somebody is going to write a new test case.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [24355]) Branching to 'twistd-umask-966'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun removed owner|
|-|-|

The branch isn't based on the attached patches; I did something simpler, skipping the whole application layer.

I expect [#3024](https://github.com/twisted/twisted/issues/3024) is also resolved by this.

Build results: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/twistd-umask-966

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_17"></a>|@washort set owner to @exarkun|
|-|-|

As mentioned above, is there really any use case for accepting non-octal umasks? Perhaps `_umask` should call `int(value, 8)` instead?

Also, 'deamonization' in `UnixApplicationRunnerSetupEnvironmentTests`' docstring and `test_deamonizedNoUmask` are misspellings.

Otherwise looks ready to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set status to `closed`|
|-|-|

(In [24393]) Merge twistd-umask-966

Author: exarkun
Reviewer: washort
Fixes: [#10647](https://github.com/twisted/twisted/issues/10647)
Fixes: [#3024](https://github.com/twisted/twisted/issues/3024)

Add a `--umask` option to `twistd` to allow the file creation mask
it sets to be customized.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
